API/ev3_inputs/ev3_input_uart.c

```diff
@@ -0,0 +1,124 @@
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/mman.h>
+#include <sys/ioctl.h>
+#include <fcntl.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <string.h>
+#include <unistd.h>
+#include "contrib/lms2012/ev3_uart.h"
+#include "include/ev3_command.h"
+#include "include/ev3_constants.h"
+#include "ev3_input_uart.h"
+
+static bool ev3UARTInputInitialized = false;
+static int uartFile = 0;
+static UART * uartSensors = NULL;
+
+extern DEVCON devCon;
+
+bool initEV3UARTInput(ANALOG * analogSensors) {
+    uartFile = open("/dev/lms_uart", O_RDWR | O_SYNC);
+    if (uartFile == -1) {
+        return false;
+    }
+    uartSensors = (UART*)mmap(0, sizeof(UART), PROT_READ | PROT_WRITE, MAP_FILE | MAP_SHARED, uartFile, 0);
+    if (uartSensors == MAP_FAILED) {
+        close(uartFile);
+        return false;
+    }
+    ev3UARTInputInitialized = true;
+
+    int i;
+    for (i = 0; i < NUM_INPUTS; i++) {
+        devCon.Connection[i] = analogSensors->InConn[i];
+        devCon.Type[i] = 0;
+        devCon.Mode[i] = 0;
+    }
+    return true;
+}
+
+
+bool setUARTSensorMode(int port, DATA8 sensorType, DATA8 sensorMode) {
+    /**
+     * Procedure to set UART mode copied from pxt: pxt-ev3/libs/core/input.ts
+     */
+    while (1) {
+        devCon.Connection[port] = CONN_INPUT_UART;
+        devCon.Type[port] = sensorType;
+        devCon.Mode[port] = sensorMode;
+
+        ioctl(uartFile, UART_SET_CONN, &devCon);
+
+        int status = waitNonZeroUARTStatusAndGet(port);
+
+        if (status & UART_PORT_CHANGED) {
+            clearUARTChanged(port);
+        } else {
+            break;
+        }
+        Wait(10);
+    }
+    return true;
+}
+
+int waitNonZeroUARTStatusAndGet(int port) {
+    while (true) {
+        int status = getUARTStatus(port);
+        if (status != 0) {
+            return status;
+        }
+        Wait(25);
+        //usleep(25000);
+    }
+}
+
+
+int getUARTStatus(int port) {
+    return uartSensors->Status[port];
+}
+
+void clearUARTChanged (int port) {
+    while (1) {
+        int status = getUARTStatus(port);
+
+        if ((status & UART_DATA_READY) != 0 && (status & UART_PORT_CHANGED) == 0) {
+            break;
+        }
+
+        devCon.Connection[port] = CONN_INPUT_UART;
+        devCon.Type[port] = 0;
+        devCon.Mode[port] = 0;
+
+        ioctl(uartFile, UART_CLEAR_CHANGED, &devCon);
```

Comment:
I think that this line contains a bug, however on the other hand everything seems to be working.

I was trying to trace how the stock lms2012 VM performs a gyro modeswitch (it somehow *does switch* the sensor out of GYRO-GnA). I was just adding some logging printf statements and then I noticed a disrepancy between this code and the kernel-side code. This IOCTL should receive a `UARTCTL` data structure and not `DEVCON`, compared to the `UART_SET_CONN`: https://github.com/mindboards/ev3sources/blob/78ebaf5b6f8fe31cc17aa5dce0f8e4916a4fc072/lms2012/d_uart/Linuxmod_AM1808/d_uart_mod.c#L3694

I'm not sure what effect it has, I will try to test it.

Reply:
Ah, I see why it works... the second line does exactly what the IOCTL does. I believe this is a bug in PXT (https://github.com/microsoft/pxt-ev3/blob/fd75bb61d6100b9d7dfb11f80d5683ea2ee398c1/libs/core/input.ts#L269). Either a correct ioctl or the and-maskout should suffice.

Reply:
Though a similar code is in the lms2012 VM as well, it for some reason does it twice (although with the correct data structure).

Reply:
Oh, this is nasty...

Somehow the memory writes from inside the kernel *do not* propagate to the userspace until after some time. Therefore the separate write removes the cached flag.

Reply:
> it somehow _does switch_ the sensor out of GYRO-GnA

Just guessing here, but this probably happens because the requested mode is not equal to the actual mode. This causes an error condition and the driver stops sending NACK messages to the sensor. The sensor times out and resets and goes through the synchronization sequence again. Then the new mode can finally be set.

Reply:
Yup, after fixing the UART code the traces seem to confirm this.
```
root@EV3:/mnt/ramdisk/prjs/test# ./test.elf 
initializing port...
MODESWITCH ON P1 TO T32 M3
ioctl; waiting for uart ready: zero ready
reading data...
MODESWITCH ON P1 TO T32 M3
ioctl; waiting for uart ready: zero ready
- rate: 0
resetting...
 [ START ] 
MODESWITCH ON P1 TO T32 M1
ioctl; waiting for uart ready: zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero zero port_changed zero zero ready
ioctl; waiting for uart ready: zero zero ready
MODESWITCH ON P1 TO T32 M3
ioctl; waiting for uart ready: zero zero ready
MODESWITCH ON P1 TO T32 M3
ioctl; waiting for uart ready: zero ready
 [ END ] 
MODESWITCH ON P1 TO T32 M3
ioctl; waiting for uart ready: zero ready
- rate: 0
```
On the `waiting for uart ready` lines a word is printed each 10 milliseconds; this equals to around 3 seconds of real time, which should be consistent with the reset timing. It also seems that by resetting the UART port by briefly configuring the port as CONN_NONE, the time can be cut to around 2.5 seconds.

Reply:
Fixes submitted to https://github.com/simonedegiacomi/EV3-API/pull/17

Reply:
> Oh, this is nasty...
> 
> Somehow the memory writes from inside the kernel _do not_ propagate to the userspace until after some time. Therefore the separate write removes the cached flag.

I'm sorry for necrobumping, but I think I have realized what is going on...

The CPU inside the EV3 brick uses a virtually-indexed-virtually-tagged cache and this type of cache should be susceptible to cache aliasing. If I understand it correctly, this means that one physical location in memory can have multiple differing values stored inside the cache if it was mapped at different virtual addresses. This is likely the case with the memory shared between the kernel and userspace. Kernel modules from LEGO use a function called remap_pfn_range() to map their memory to userspace and that indeed creates a new independent mapping from virtual to physical address (https://stackoverflow.com/a/8788593). Therefore I think that while the `ioctl()` call and the write to the `mmap`'ed status variable access the same physical memory location, they access a different cache entry and so both are needed.

I don't yet understand how the conflict is resolved, but I would think that both cache entries get periodically flushed by other tasks preempting the current program.